utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +697,466 @@ impl FromStr for FixedDecimal {
+                let sig = sig as usize;
+                if sig == 0 {
+                    return Err(Error::Limit);
+                }
+                if sig < n_digits {
+                    let round_by = (n_digits - sig) as u16;
+                    decimal.round_trailing_digits(round_by, mode)?;
+                    // It may have rounded up by one
+                    debug_assert!(decimal.digits.len() <= sig as usize);
+                }
+                let target_lowest_magnitude = decimal.magnitude - sig as i16 + 1;
+                if target_lowest_magnitude <= 0 {
+                    // If the target magnitude was negative, make sure we update
+                    // the lower magnitude to match it
+                    decimal.lower_magnitude = target_lowest_magnitude;
+                } else if decimal.lower_magnitude < 0 {
+                    // If the target magnitude was positive,
+                    // we may have truncated digits off the right side
+                    // of the decimal point, reset the lower magnitude
+                    decimal.lower_magnitude = 0;
+                }
+            }
+        }
+        Ok(decimal)
+    }
+    /// Internal function for parsing directly from floats using ryū
+    fn new_from_f64_raw(float: f64) -> Result<Self, Error> {
+        if !float.is_finite() {
+            return Err(Error::Limit);
+        }
+        // note: this does not heap allocate
+        let mut buf = ryu::Buffer::new();
+        let formatted = buf.format_finite(float);
+        Self::from_str(formatted)
+    }
+
+    /// Internal function to round off `n` digits
+    /// from the right
+    ///
+    /// `self` must have at least `n` digits
+    ///
+    /// This may end up adding a digit to the left!
+    ///
+    /// This will not change the number of significant digits, it simply exists
+    /// to *round* them (and will typically reduce the size of `self.digits`)
+    ///
+    /// This function is responsible for fixing `digits`, `magnitude`, and `upper_magnitude`.
+    /// It will only modify upper_magnitude when it is not large enough to fit the rounded number.
+    /// The caller may fix up `lower_magnitude` by whatever scheme it desires
+    fn round_trailing_digits(&mut self, n: u16, mode: RoundingMode) -> Result<(), Error> {
+        debug_assert!(
+            self.digits.len() >= n as usize,
+            "Attempted to round off {} digits of number that has only {}",
+            n,
+            self.digits.len()
+        );
+        if n == 0 {
+            // no point attempting to round off any digits
+            return Ok(());
+        }
+
+        let cutoff = self.digits.len() - n as usize;
+
+        match mode {
+            RoundingMode::Unnecessary => {
+                // If we got to this point then rounding was not unnecessary
+                return Err(Error::Limit);
+            }
+            RoundingMode::Truncate => {
+                self.digits.truncate(cutoff);
```

Comment:
Issue: you can still end up with trailing zeros here ("1.02" => "1.0").

Please fix, add a unit test, and then add, before every return statement in this function,

```rust
        #[cfg(debug_assertions)]
        result.check_invariants();
```